packages/store/src/external/hydrate.js

```diff
@@ -0,0 +1,115 @@
+// @ts-check
+
+import '../types';
+
+import { makeWeakStore } from '../weak-store';
+import { makeStore } from '../store';
+
+/**
+ * @callback MakeBackingStore
+ * @param {HydrateHook} hydrateHook
+ * @returns {BackingStore}
+ */
+
+/**
+ * This creates an external store maker for a given storage backend, supporting
+ * the Closure interface that the rewriter targets.
+ *
+ * @template {Array<any>} A
+ * @template {ExternalInstance} T
+ * @param {MakeBackingStore} makeBackingStore
+ * @returns {MakeHydrateExternalStore<A, T>}
+ */
+export const makeHydrateExternalStoreMaker = makeBackingStore => {
+  const serialize = JSON.stringify;
+  const unserialize = JSON.parse;
+
+  /** @type {WeakStore<T, [string, string]>} */
+  const instanceToKey = makeWeakStore('instance');
+
+  let lastStoreKey = 0;
+
+  // This has to be a strong store, since it is indexed by key.
+  const storeKeyToHydrate = makeStore('storeKey');
+
+  /**
+   * Create a data object that queues writes to the store.
+   *
+   * @param {HydrateData} data
+   * @param {() => void} markDirty
+   */
+  const makeActiveData = (data, markDirty) => {
+    const activeData = {};
+    // For every property in data...
+    for (const prop of Object.getOwnPropertyNames(data)) {
+      // Define a getter and setter on activeData.
+      Object.defineProperty(activeData, prop, {
+        get: () => data[prop],
+        set: value => {
+          data[prop] = value;
+          markDirty();
+        },
+      });
+    }
+    return harden(activeData);
+  };
+
+  /**
+   * @type {BackingStore}
+   */
+  let backing;
+  const hydrateHook = {
+    getKey(value) {
+      return instanceToKey.get(value);
+    },
+    load([storeKey, instanceKey]) {
+      const hydrate = storeKeyToHydrate.get(storeKey);
+      const store = backing.findStore(storeKey);
+
+      const data = unserialize(store.get(instanceKey));
+      const markDirty = () => store.set(instanceKey, serialize(data));
+
+      const activeData = makeActiveData(data, markDirty);
+      const obj = hydrate(activeData);
+      instanceToKey.init(obj, [storeKey, instanceKey]);
+      return obj;
+    },
+    drop(storeKey) {
+      storeKeyToHydrate.delete(storeKey);
+    },
+  };
+
+  backing = makeBackingStore(hydrateHook);
+
+  function makeHydrateExternalStore(instanceName, adaptArguments, makeHydrate) {
```

Comment:
The term "instanceName" confused me. In this scheme, when I hear "instance", I immediately think about the objects made by the maker, which are the arguments of `instanceToKey`. Can we call this something like `storeName`?

Reply:
> This system is so beautiful, and solves such a crucial problem for us so simply, that I actually teared up reading this code. Really. I can't tell you how much I appreciate seeing this problem collapse into this solution. Thanks!

Thank you!  :blush: I'm really glad you insisted on exploring rewrites as an avenue.  The work is far from done, but we're getting there.

Reply:
> The term "instanceName" confused me. [...] Can we call this something like `storeName`?

I moved back to 'keyName', which is the same thing we used for the `makeWeakStore`/`makeStore` argument.  If you'd like to change that, we should change all three places.

Reply:
I still see `instanceName`. Did you forget to push? Unresolving.

Reply:
Oops.  Fixed. :smile: